src/trusted/service_runtime/sel_ldr.c

```diff
@@ -1595,14 +1529,89 @@ void NaClCopyDynamicText(struct NaClApp *nap_parent, struct NaClApp *nap_child)
 
   /* copy dynamic text regions */
   NaClXMutexLock(&nap_child->dynamic_load_mutex);
-  /* NaClDyncodeVisit() handles parent dynamic_load_mutex lock */
+  NaClXMutexLock(&nap_parent->dynamic_load_mutex);
   nap_child->text_shm = nap_parent->text_shm;
-  NaClDyncodeVisit(nap_parent, NaClCopyDynamicRegion, nap_child);
+  for (ssize_t i = 0; i < nap_parent->num_dynamic_regions; ++i) {
+    NaClCopyDynamicRegion(nap_child, &nap_parent->dynamic_regions[i]);
+  }
+  NaClXMutexUnlock(&nap_parent->dynamic_load_mutex);
   NaClXMutexUnlock(&nap_child->dynamic_load_mutex);
 
   /* copy page mappings */
-  NaClVmmapVisit(&nap_parent->mem_map, NaClVmCopyEntry, nap_child);
 
+  NaClVmmapMakeSorted(parentmap);
+  for (size_t i = 0, nentries = parentmap->nvalid; i < nentries;) {
+    short iters = nentries - i < IOV_MAX ? nentries - i : IOV_MAX;
+
+    for(int iters1 = 0; iters1 < iters; ++iters1, ++i) {
+      struct NaClVmmapEntry *entry = parentmap->vmentry[i];
+      uintptr_t page_addr_child = (entry->page_num << NACL_PAGESHIFT) | offset;
+      uintptr_t page_addr_parent = (entry->page_num << NACL_PAGESHIFT) | parent_offset;
+      size_t copy_size = entry->npages << NACL_PAGESHIFT;
+
+      NaClLog(2, "copying %zu page(s) at %zu [%#lx] from (%p) to (%p)\n",
+              entry->npages,
+              entry->page_num,
+              copy_size,
+              (void *)page_addr_parent,
+              (void *)page_addr_child);
+      if(entry->prot && (entry->desc != nap_parent->text_shm)) {
+        if (NaClMprotect((void *)page_addr_parent, copy_size, entry->prot) == -1) {
+          NaClLog(LOG_FATAL, "%s\n", "parent vmmap page NaClMprotect failed!");
+        }
+        if (NaClMprotect((void *)page_addr_child, copy_size, PROT_RW) == -1) {
+          NaClLog(LOG_FATAL, "%s\n", "child vmmap page NaClMprotect failed!");
+        }
+        inputvector[veccount].iov_base = (void*) page_addr_parent;
+        outputvector[veccount].iov_base = (void*) page_addr_child;
+        inputvector[veccount].iov_len = copy_size;
+        outputvector[veccount].iov_len = copy_size;
+        ++veccount;
+      }
+    }
+    i -= iters;
+
+    {
```

Comment:
Extra brackets?

Reply:
The brackets are there for 2 reasons:
1. To create a separate scope for the iteration to happen in so that it's clear this is it's own thing not part of the other logic for the main loop
2. Mostly to avoid compiler warnings about ANSI C noncompliance because of mixed declarations and code